### PR TITLE
server: support mysql CLIENT_CONNECT_ATTRS capability

### PR DESCRIPTION
**Pull request by tiancaiamao**

### CODE REVIEW

**Comment by shenli** [comment]
Any test case?

**Comment by tiancaiamao** [comment]
test add @shenli

File: server/conn.go [hunk 1]
```diff
@@ -163,38 +165,113 @@ func (cc *clientConn) writePacket(data []byte) error {
...
-	authLen := int(data[pos])
-	pos++
-	auth := data[pos : pos+authLen]
-	pos += authLen
-	if cc.capability&mysql.ClientConnectWithDB > 0 {
+	packet.User = string(data[pos : pos+bytes.IndexByte(data[pos:], 0)])
+	pos += len(packet.User) + 1
+
+	if capability&mysql.ClientPluginAuthLenencClientData > 0 {
+		// TODO support mysql.ClientPluginAuthLenencClientData, skip it now
```
Comment (zimulala) [inline server/conn.go]: `TODO` -> `TODO:`, the same as line 219.

File: server/conn.go [hunk 2]
```diff
@@ -163,38 +165,113 @@ func (cc *clientConn) writePacket(data []byte) error {
...
 		}
 	}
+	return nil
+}
+
+func parseAttrs(data []byte) (map[string]string, error) {
+	attrs := make(map[string]string)
+	pos := 0
+	for pos < len(data) {
+		key, _, off, err := parseLengthEncodedBytes(data[pos:])
```
Comment (zimulala) [inline server/conn.go]: Why not check `isNull`？

Comment (tiancaiamao) [inline server/conn.go]: if `isNull` is true, key would be an empty slice, it doesn't make any trouble but save several line of code.

File: server/conn.go [hunk 3]
```diff
@@ -163,38 +165,113 @@ func (cc *clientConn) writePacket(data []byte) error {
 	return cc.pkg.writePacket(data)
 }
 
-func (cc *clientConn) readHandshakeResponse() error {
-	data, err := cc.readPacket()
-	if err != nil {
-		return errors.Trace(err)
-	}
+type HandshakeResponse41 struct {
```
Comment (shenli) [inline server/conn.go]: Why make this public?

File: server/conn.go (hunk 2 above)
Comment (zimulala) [inline server/conn.go]: But the `attrs` will produce dirty data when it sets `attrs` with empty `key` .

Comment (tiancaiamao) [inline server/conn.go]: No, that's not dirty data.

Let me explain how mysql encode string first: 

string can be encoded with [length:data], where length is something similar to varint encoding.
how to encode a `NULL` string? no data need and length be encoded as a special flag, that's exactly "" in Go, and the case when  `isNull` return true.

so, when we meet `isNull` in the key, I think user says that he want specify an `attrs` with ("" , value).

**Comment by zimulala** [comment]
LGTM

**Comment by shenli** [comment]
LGTM
Please make sure all tidb tests is past.

**Pull request merged**